src/ledger/LedgerStateSnapshot.h

```diff
@@ -0,0 +1,161 @@
+    virtual LedgerEntryWrapper getAccount(LedgerHeaderWrapper const& header,
+                                          TransactionFrame const& tx,
+                                          AccountID const& AccountID) const = 0;
+    virtual LedgerEntryWrapper load(LedgerKey const& key) const = 0;
+    virtual void executeWithMaybeInnerSnapshot(
+        std::function<void(LedgerSnapshot&)> f) const = 0;
+};
+
+// A concrete implementation of read-only SQL snapshot wrapper
+class LedgerTxnReadOnly : public AbstractLedgerStateSnapshot
+{
+    // Callers are expected to manage `AbstractLedgerTxn` themselves.
+    // LedgerTxnReadOnly guarantees that LedgerTxn, LedgerTxnHeader and
+    // LedgerTxnEntry referenced all remain valid.
+    AbstractLedgerTxn& mLedgerTxn;
+
+  public:
+    LedgerTxnReadOnly(AbstractLedgerTxn& ltx);
+    ~LedgerTxnReadOnly() override;
+    LedgerHeaderWrapper getLedgerHeader() const override;
+    LedgerHeaderWrapper getLedgerHeaderUnsafe() const override;
+    LedgerEntryWrapper getAccount(AccountID const& account) const override;
+    LedgerEntryWrapper getAccount(LedgerHeaderWrapper const& header,
+                                  TransactionFrame const& tx) const override;
+    LedgerEntryWrapper getAccount(LedgerHeaderWrapper const& header,
+                                  TransactionFrame const& tx,
+                                  AccountID const& AccountID) const override;
+    LedgerEntryWrapper load(LedgerKey const& key) const override;
+    void executeWithMaybeInnerSnapshot(
+        std::function<void(LedgerSnapshot&)> f) const override;
+};
+
+// A concrete implementation of read-only BucketList snapshot wrapper
+class BucketSnapshotState : public AbstractLedgerStateSnapshot
+{
+    std::shared_ptr<SearchableBucketListSnapshot> mSnapshot;
+
+  public:
+    BucketSnapshotState(BucketSnapshotManager& bsm);
+    ~BucketSnapshotState() override;
+
+    LedgerHeaderWrapper getLedgerHeader() const override;
+    LedgerHeaderWrapper getLedgerHeaderUnsafe() const override;
+    LedgerEntryWrapper getAccount(AccountID const& account) const override;
+    LedgerEntryWrapper getAccount(LedgerHeaderWrapper const& header,
+                                  TransactionFrame const& tx) const override;
+    LedgerEntryWrapper getAccount(LedgerHeaderWrapper const& header,
+                                  TransactionFrame const& tx,
+                                  AccountID const& AccountID) const override;
+    LedgerEntryWrapper load(LedgerKey const& key) const override;
+    void executeWithMaybeInnerSnapshot(
+        std::function<void(LedgerSnapshot&)> f) const override;
+};
+
+// A helper class to create and query read-only snapshots
+// Automatically decides whether to create a BucketList (recommended), or SQL
+// snapshot (deprecated, but currently supported)
+class LedgerSnapshot : public NonMovableOrCopyable
+{
+    std::shared_ptr<AbstractLedgerStateSnapshot> mGetter;
```

Comment:
Could these point at const values (and also probably be const themselves)?

Reply:
made the value const, but the kept the pointer itself non-const (the constructor is non-trivial, so the refactor of it looked a bit awkward)